### PR TITLE
Tcp reconnect

### PR DESCRIPTION
**Pull request by edersondisouza**
#32

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
I don't really like the use of std::function because it doesn't have a constexpr constructor... meaning there will be an allocated object per call. Or am I not up to date on its usage?

**Comment by edersondisouza** [comment]
> I don't really like the use of std::function because it doesn't have a constexpr constructor... meaning there will be an allocated object per call. Or am I not up to date on its usage?

Glorious C++ and infinite different (or new to me) information to consider...

Maybe did you mean that for `std::bind`? We are going to have one `std::function` per `Timeout` - nothing really different from what we do have now.
If you are worried about many `std::bind` calls, them we might consider having it as field of every log object - not sure if worth, as it's not a code path that I would expect to have many runs: two binds for each start.
Or maybe you meant something else...

**Comment by lucasdemarchi** [comment]
Yep, std::bind. Context:

~2 years ago I wanted to use std::function on ardupilot to replace its awful "fastdelegate" implementation. I stopped pursuing that due to: 

- There's no way to convince stdc++ to create objects at compile time. You always need an extra allocation  - this is due to the missing constexpr constructor. In ArduPilot this was really critical because then the object would not stay in flash
- Some benchmarks (on Linux) showed it to be slower and you can find some benchmarks available on the web... This may not be that important in the current use case, but we may need to think if this will continue to be true in future... and it's mind-blowing since the only thing we really want is a callback with a data pointer to be called in future, wth would we need the extra allocation and be so slow?

So... I ended up in a slightly different/simpler functor implementation for ardupilot:
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL/utility/functor.h

I'm not saying to use it. I'm all in favor of user the standard interfaces if they cover our use cases (e.g. we could make more use here of std::string, std::unique_ptr, etc). In order to make such decisions it's nice to take a look on what other projects are doing... I always use chromium/webkit as a reference for c++ when it makes sense. See for example this discussion: https://groups.google.com/a/chromium.org/forum/#!topic/cxx/SoEj7oIDNuA

File: endpoint.h [hunk 1]
```diff
@@ -145,15 +145,30 @@ class UdpEndpoint : public Endpoint {
...
     int flush_pending_msgs() override { return -ENOSYS; }
 
     struct sockaddr_in sockaddr;
+    int retry_timeout = 0;
+
+    inline const char *get_ip() {
+        return _ip;
+    }
+
+    inline unsigned long get_port() {
```
Comment (lucasdemarchi) [inline endpoint.h]: it's implicitly inline when definition is on the header

**Comment by fnoop** [comment]
New parameter should be added to Readme.
So it looks like the tcp endpoint does reconnect if connection lost, but the routing seems to break.  After reconnection I can no longer connect to SITL (the tcp endpoint) through another endpoint.

**Comment by edersondisouza** [comment]
@fnoop that's sad... I'll look into it

**Comment by edersondisouza** [comment]
@fnoop #50 should fix it.

**Pull request merged**